src/core/dense_set.cc

```diff
@@ -0,0 +1,424 @@
+  }
+
+  // otherwise make a new link and connect it to the front of the list
+  LinkAllocator la(mr());
+  DenseLinkKey* lk = la.allocate(1);
+  la.construct(lk);
+
+  lk->next = *it;
+  it->SetLink(lk);
+  lk->SetObject(data);
+
+  return ObjectAllocSize(data);
+}
+
+void DenseSet::PushFront(DenseSet::ChainVectorIterator it, DenseSet::DensePtr ptr) {
+  if (it->IsEmpty()) {
+    it->SetObject(ptr.GetObject());
+    if (ptr.IsLink()) {
+      // deallocate the link if it is no longer a link as it is now in an empty list
+      mr()->deallocate(ptr.Raw(), sizeof(DenseLinkKey), alignof(DenseLinkKey));
+    }
+  } else if (ptr.IsLink()) {
+    // if the pointer is already a link then no allocation needed
+    *ptr.Next() = *it;
+    *it = ptr;
+  } else {
+    // allocate a new link if needed and copy the pointer to the new link
+    LinkAllocator la(mr());
+    DenseLinkKey* lk = la.allocate(1);
+    la.construct(lk);
+
+    lk->next = *it;
+    lk->SetObject(ptr.GetObject());
+    it->SetLink(lk);
+  }
+}
+
+auto DenseSet::PopPtrFront(DenseSet::ChainVectorIterator it) -> DensePtr {
+  if (it->IsEmpty()) {
+    return DensePtr(nullptr);
+  }
+
+  DensePtr front = *it;
+
+  // if this is an object, then it's also the only record in this chain.
+  // therefore, we should just reset DensePtr.
+  if (it->IsObject()) {
+    it->Reset();
+  } else {
+    DCHECK(it->IsLink());
+    // this will not return nullptr since it is a Link
+    *it = *it->Next();
+  }
+
+  return front;
+}
+
+void* DenseSet::PopDataFront(DenseSet::ChainVectorIterator it) {
+  DensePtr front = PopPtrFront(it);
+  void* ret = front.GetObject();
+
+  if (front.IsLink()) {
+    DenseLinkKey* lk = (DenseLinkKey*)front.Raw();
+    mr()->deallocate(lk, sizeof(DenseLinkKey), alignof(DenseLinkKey));
+  }
+
+  return ret;
+}
+
+auto DenseSet::Unlink(DenseSet::DensePtr* node) -> DensePtr {
+  DensePtr ret = *node;
+  if (node->IsObject()) {
+    node->Reset();
+  } else {
+    *node = *node->Next();
+  }
+
+  return ret;
+}
+
+void* DenseSet::UnlinkAndFree(DenseSet::DensePtr* node) {
+  DensePtr unlinked = Unlink(node);
+  void* ret = unlinked.GetObject();
+
+  if (unlinked.IsLink()) {
+    mr()->deallocate(unlinked.Raw(), sizeof(DenseLinkKey), alignof(DenseLinkKey));
+  }
+
+  return ret;
+}
+
+void DenseSet::Clear() {
+  for (auto it = entries_.begin(); it != entries_.end(); ++it) {
+    while (!it->IsEmpty()) {
+      PopDataFront(it);
+    }
+  }
+
+  entries_.clear();
+}
+
+auto DenseSet::FindEmptyAround(uint32_t bid) -> ChainVectorIterator {
+  if (entries_[bid].IsEmpty()) {
+    return entries_.begin() + bid;
+  }
+
+  if (!kAllowDisplacements) {
+    return entries_.end();
+  }
+
+  if (bid + 1 < entries_.size() && entries_[bid + 1].IsEmpty()) {
+    return entries_.begin() + bid + 1;
+  }
+
+  if (bid && entries_[bid - 1].IsEmpty()) {
+    return entries_.begin() + bid - 1;
+  }
+
+  return entries_.end();
+}
+
+bool DenseSet::Reserve(size_t sz) {
```

Comment:
why did you feel you need to check for success in this function? i.e. why check if `entries_.capacity() != sz` and if 
`sz > entries_.max_size()` ?

Reply:
That was originally added in the case of checking if the vector ran into reallocation issues